test/e2e/framework/kcp.go

```diff
@@ -207,6 +211,164 @@ func GatherMetrics(ctx context.Context, t *testing.T, server RunningServer, dire
+	labels := map[string]string{
+		"server": srv.Name(),
+		"test":   t.Name(),
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), wait.ForeverTestTimeout)
+	defer cancel()
+	require.NoError(t, ScrapeMetrics(ctx, srv.RootShardSystemMasterBaseConfig(t), promUrl, promCfgDir, jobName, labels))
+}
+
+func ScrapeMetrics(ctx context.Context, cfg *rest.Config, promUrl, promCfgDir, jobName string, labels map[string]string) error {
+	type staticConfigs struct {
+		Targets []string          `yaml:"targets,omitempty"`
+		Labels  map[string]string `yaml:"labels,omitempty"`
+	}
+	type tlsConfig struct {
+		InsecureSkipVerify bool `yaml:"insecure_skip_verify,omitempty"`
+	}
+	type scrapeConfig struct {
+		JobName        string          `yaml:"job_name,omitempty"`
+		ScrapeInterval string          `yaml:"scrape_interval,omitempty"`
+		BearerToken    string          `yaml:"bearer_token,omitempty"`
+		TlsConfig      tlsConfig       `yaml:"tls_config,omitempty"`
+		Scheme         string          `yaml:"scheme,omitempty"`
+		StaticConfigs  []staticConfigs `yaml:"static_configs,omitempty"`
+	}
+	type config struct {
+		ScrapeConfigs []scrapeConfig `yaml:"scrape_configs,omitempty"`
+	}
+	promCfg := config{}
+
+	scrapeConfigFile := filepath.Join(promCfgDir, "prom.yaml")
+	f, err := os.OpenFile(scrapeConfigFile, os.O_RDWR|os.O_CREATE, 0o644)
+	if err != nil {
+		return err
+	}
+
+	// lock config file exclusively, blocks all other producers until unlocked or process (test) exits
+	err = syscall.Flock(int(f.Fd()), syscall.LOCK_EX)
+	if err != nil {
+		return err
+	}
+	err = gopkgyaml.NewDecoder(f).Decode(&promCfg)
+	if err != nil && !errors.Is(err, io.EOF) {
+		return err
+	}
+
+	hostUrl, err := url.Parse(cfg.Host)
+	if err != nil {
+		return err
+	}
+	promCfg.ScrapeConfigs = append(promCfg.ScrapeConfigs, scrapeConfig{
+		JobName:        jobName,
+		ScrapeInterval: (5 * time.Second).String(),
+		BearerToken:    cfg.BearerToken,
+		TlsConfig:      tlsConfig{InsecureSkipVerify: true},
+		Scheme:         hostUrl.Scheme,
+		StaticConfigs: []staticConfigs{{
+			Targets: []string{hostUrl.Host},
+			Labels:  labels,
+		}},
+	})
+	err = f.Truncate(0)
+	if err != nil {
+		return err
+	}
+	_, err = f.Seek(0, 0)
+	if err != nil {
+		return err
+	}
+	err = gopkgyaml.NewEncoder(f).Encode(&promCfg)
+	if err != nil {
+		return err
+	}
+	err = syscall.Flock(int(f.Fd()), syscall.LOCK_UN)
+	if err != nil {
+		return err
+	}
+	f.Close()
+
+	_, err = http.Post(promUrl+"/-/reload", "", nil)
+	if err != nil {
+		return err
+	}
+
+	type targets struct {
+		Data struct {
+			ActiveTargets []struct {
+				Health     string `json:"health,omitempty"`
+				ScrapePool string `json:"scrapePool,omitempty"`
+			} `json:"activeTargets,omitempty"`
+		} `json:"data,omitempty"`
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	err = nil
+	go func() {
+		i := 0
+		defer wg.Done()
+		for {
+			err = func() error {
+				b, err := http.Get(promUrl + "/api/v1/targets")
```

Comment:
side note: waiting for active target availability is not strictly needed, however I found that e2e tests sometimes run so quickly that Prometheus might miss to scrape them initially.